### PR TITLE
Add numbered variant to code snippet

### LINKED ISSUE #3471: Implement line numbering for code snippet

### PR DESCRIPTION
**Pull request by sowasred2012**
## Done

Added numbered line variant of code snippet block.

Fixes #3471 
## QA

- Open [demo](https://vanilla-framework-3490.demos.haus/docs/examples/patterns/code-snippet/code-snippet-numbered)
- See that it matches the [design](https://app.zeplin.io/project/5fa270d8eee4dbb267e579f1/screen/5fc616c95320fdb32885432e)
- Review updated documentation:
  - Check the updated [documentation](https://vanilla-framework-3490.demos.haus/docs/base/code) makes sense

### CODE REVIEW

File: templates/docs/base/code.md [hunk 1]
```diff
@@ -69,27 +77,13 @@ View example of the code copyable pattern
...
-The code numbered pattern can be used when displaying large blocks of code to enable users to quickly reference a specific line.
-
-<div class="embedded-example"><a href="/docs/examples/patterns/code-numbered/" class="js-example">
-View example of the code numbered pattern
-</a></div>
-
-### Import
-
-To import just code copyable or numbered component into your project, copy either or both snippets below and include it in your main Sass file.
+### Deprecated numbered
```
Comment (bartaz) [inline templates/docs/base/code.md:80]: Keep the title, by put `<span class="p-label--deprecated">Deprecated</span>` below.

Comment (bartaz) [reply templates/docs/base/code.md:80]: Ah, you have the label. So keep the title

File: templates/docs/base/code.md [hunk 2]
```diff
@@ -38,10 +38,18 @@ View example of the base code block
```
Comment (bartaz) [inline templates/docs/base/code.md:38]: Just noticed ^ I don't think we use plurals and title case in headings, so probably should be "Code snippet"

File: templates/docs/examples/patterns/code-snippet/code-snippet-numbered.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Code Snippet / Numbered{% endblock %}
```
Comment (bartaz) [inline templates/docs/examples/patterns/code-snippet/code-snippet-numbered.html:2]: "Code snippet" in the example names as well for consistency (same in other code snippet examples).

File: scss/_patterns_code-numbered.scss [hunk 4]
```diff
@@ -2,43 +2,45 @@
...
-      text-align: right;
-      top: 0;
-      // stylelint-disable property-no-vendor-prefix
-      -webkit-user-select: none;
-      -moz-user-select: none;
-      -ms-user-select: none;
-      user-select: none;
-      // stylelint-enable property-no-vendor-prefix
-      width: $sidebar-width;
+      padding: 0 $sph-inner 0 ($sidebar-width + $sph-inner);
```
Comment (bartaz) [inline scss/_patterns_code-numbered.scss:13]: Would be nice to get rid of that amount of duplication, as it may be hard to keep them in line if any changes are made.

I'm not exactly sure where to put common parts. We don't want the legacy 'code numbered' to depend on mixins or placeholders from new pattern. Doing it other way around and making new code snippet take styles from legacy component doesn't seem good as well.

Maybe we should have some `%code-numbered` defined in `_base_code.scss`, and extended where needed?

File: templates/docs/base/code.md [hunk 5]
```diff
@@ -81,25 +87,11 @@ Please copy the entire JS in the example, for copy to clipboard functionality.
...
-</a></div>
-
-### Import
-
-To import just code copyable or numbered component into your project, copy either or both snippets below and include it in your main Sass file.
-
-```scss
-@import 'patterns_code-copyable';
-```
+<span class="p-label--deprecated">Deprecated</span>
```
Comment (bartaz) [inline templates/docs/base/code.md:90]: Please add the status change to the 'components status' table as well.

File: scss/_patterns_code-numbered.scss (hunk 4 above)
Comment (sowasred2012) [reply scss/_patterns_code-numbered.scss:13]: I was in two minds about this - I thought about creating a placeholder, but once this pattern is removed, the only place it'll be used is in the code snippet, defeating the purpose of a placeholder. 

As we're deprecating the older pattern, and it'll ultimately get removed, I thought duplication wasn't so bad in this case and that the older pattern could potentially stay "frozen" this way until it is removed.

Comment (bartaz) [reply scss/_patterns_code-numbered.scss:13]: While we are deprecating it, it's not like we are going to remove it soon, and we want the style consistent, so I don't think we want to keep 2 separate implementations.

Yes, using placeholder means it will technically not be needed later, when there is one component only. But it also doesn't hurt.

I wouldn't mind if it's just a bit of CSS, but there is quite a lot of very specific styling here which we may want to keep up to date properly.

**Comment by bartaz** [comment]
The spacing of the numbers is quite large (even for large numbers), maybe we should reduce it as @lyubomir-popov suggested?

Screenshot 2021-01-14 at 13 39 10

**Comment by sowasred2012** [comment]
@lyubomir-popov could you take a look at the width of the number column? I've reduced it some, it should now fit 4 numbers comfortably whilst occupying less space.

**Comment by lyubomir-popov** [comment]
@sowasred2012 we can safely shave off another .5rem. Since the font is monospaced, all digits occupy the same space. That seems to be almost exactly .5rem per char:
image

So for 4 digits we need 2rem, and we have one rem padding on each side, so 4rem total:

image

File: scss/_patterns_code-numbered.scss [hunk 6]
```diff
@@ -2,43 +2,45 @@
 $sidebar-width: 4.5rem !default;
```
Comment (lyubomir-popov) [inline scss/_patterns_code-numbered.scss:2]:
```suggestion
$digit-width: .5rem; // measured width of one character in the monospaced font
$sidebar-width: $sph-inner + 4 * $digit-width + $sph-inner !default;
```

**Comment by lyubomir-popov** [comment]
@sowasred2012 maybe add `overflow: hidden;` to be safe? couldn't get `text-overflow:ellipsis` to work though.

File: scss/_patterns_code-numbered.scss [hunk 7]
```diff
@@ -1,44 +1,20 @@
...
-      -webkit-user-select: none;
-      -moz-user-select: none;
-      -ms-user-select: none;
-      user-select: none;
-      // stylelint-enable property-no-vendor-prefix
-      width: $sidebar-width;
+    .p-code-numbered__line {
+      @extend %code-numbered-line;
+
+      &::before {
```
Comment (bartaz) [inline scss/_patterns_code-numbered.scss:14]: Why is this needed?

File: scss/_patterns_code-numbered.scss [hunk 8]
```diff
@@ -1,44 +1,20 @@
...
-    white-space: pre-wrap;
-    width: 100%;
-
-    &:empty {
-      display: block;
-      min-height: $sp-xx-large;
+  @include deprecate('3.0.0', 'p-code-numbered is being removed, use the code snippet block numbered variant instead.') {
+    .p-code-numbered {
+      counter-reset: line-numbering;
+      padding: $sph-inner--small 0;
```
Comment (bartaz) [inline scss/_patterns_code-numbered.scss:8]: Shouldn't code numbered have same padding as default code block? we don't need to 0 the padding for the white background anymore.

File: scss/_patterns_code-snippet.scss [hunk 9]
```diff
@@ -44,6 +45,15 @@ $color-snippet-heading-bg: rgba($color-x-dark, 0.08);
       }
     }
 
+    .p-code-snippet__block--numbered {
+      counter-reset: line-numbering;
+      padding: $sph-inner--small 0;
```
Comment (bartaz) [inline scss/_patterns_code-snippet.scss:50]: Same as with legacy component - does it need to override the padding? I guess we can accomodate the padding on the lines with padding on the block, and avoid overriding this.

**Comment by bartaz** [comment]
@lyubomir-popov Would that be really bad if (in the edge case of exceeding 10000 lines) we let the fifth digit overflow on the left padding (instead of just cropping the last one)?

Screenshot 2021-01-15 at 08 58 30

Basically it's doing `left: -0.5rem` on the number ::before, and bumping its sizes by 0.5rem (up to 5 digits).

Sure, it doesn't look great, but it doesn't break the numbering. And should likely be very very rare (I can't imagine having a code snippet on a page that long, even on things like github such long files should be rare).

File: scss/_patterns_code-numbered.scss [hunk 10]
```diff
@@ -1,44 +1,14 @@
 @import 'settings';
-$sidebar-width: 4.5rem !default;
+$sidebar-width: 3rem !default;
```
Comment (bartaz) [inline scss/_patterns_code-numbered.scss:2]: we already have `$code-sidebar-width` defined, is it different?

Comment (bartaz) [reply scss/_patterns_code-numbered.scss:2]: Actually, this one doesn't seem to be used at all ATM.

File: templates/docs/base/code.md [hunk 11]
```diff
@@ -81,25 +87,11 @@ Please copy the entire JS in the example, for copy to clipboard functionality.
...
 
-```scss
-@import 'patterns_code-numbered';
-```
+The code numbered pattern is now deprecated and will be removed in a future version of Vanilla.
 
-For more information see [Customising Vanilla](/docs/customising-vanilla/) in your projects, which includes overrides and importing instructions.
+It is being replaced by the new [code snippet pattern](/docs/base/code#code-snippets).
```
Comment (bartaz) [inline templates/docs/base/code.md:94]: This link doesn't work, the heading is not plural
```suggestion
It is being replaced by the new [code snippet pattern](/docs/base/code#code-snippet).
```

File: templates/docs/component-status.md [hunk 12]
```diff
@@ -29,6 +29,12 @@ When we add, make significant updates, or deprecate a component we update their
       <td>2.22.0</td>
       <td>The new <code>p-separator</code> component has been added to be used as a separator between content blocks.</td>
     </tr>
+    <tr>
```
Comment (bartaz) [inline templates/docs/component-status.md:32]: When deprecating and suggesting new component, we should add info about Code snippet being added as well.

**Comment by lyubomir-popov** [comment]
> Would that be really bad if (in the edge case of exceeding 10000 lines) we let the fifth digit overflow on the left padding (instead of just cropping the last one)?

looks ok to me

File: scss/_base_code.scss [hunk 13]
```diff
@@ -1,6 +1,8 @@
 @import 'settings';
 
 $color-pre-bg: rgba($color-x-dark, 0.03);
+$digit-width: 0.5rem; // measured width of one character in the monospaced font
```
Comment (anthonydillon) [inline scss/_base_code.scss:4]: Could we use `1ch` here?

**Comment by bartaz** [comment]
@sowasred2012 There is something going wrong when there are empty lines in code:

Screenshot 2021-01-15 at 10 17 40

File: scss/_base_code.scss (hunk 13 above)
Comment (sowasred2012) [reply scss/_base_code.scss:4]: TIL about `ch`... however, unfortunately not, in the next line we're combining `$digit-width` with a `rem` value, `$sph-inner`, and `ch` and `rem` values are incompatible.

Comment (lyubomir-popov) [reply scss/_base_code.scss:4]: that's what calc is for e.g. `calc(2rem + 4cm)` :)

Comment (sowasred2012) [reply scss/_base_code.scss:4]: Ah yes of course! Cool, adding this in now.

File: scss/_base_code.scss [hunk 14]
```diff
@@ -1,5 +1,9 @@
 @import 'settings';
 
+$color-pre-bg: rgba($color-x-dark, 0.03);
+$digit-width: 1ch; // measured width of one character in the monospaced font
+$code-sidebar-width: calc(#{$sph-inner} + (4 * #{$digit-width})) !default;
```
Comment (bartaz) [inline scss/_base_code.scss:5]: if the 'digit width' uses `1ch` (which literary is 1 character width) I don't think we need separate variable for that.
```suggestion
$code-sidebar-width: calc(#{$sph-inner} + 4ch) !default; // width based on padding + 4 digits width
```

File: templates/docs/examples/patterns/code-snippet/code-snippet-numbered.html [hunk 15]
```diff
@@ -0,0 +1,23 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Code snippet / Numbered{% endblock %}
+
+{% block standalone_css %}patterns_code-snippet{% endblock %}
+
+{% block content %}
+<div class="p-code-snippet">
+  <div class="p-code-snippet__header">
+    <h5 class="p-code-snippet__title">Check you have snap installed</h5>
```
Comment (bartaz) [inline templates/docs/examples/patterns/code-snippet/code-snippet-numbered.html:9]: The header doesn't seems to be correct for the code content. Update it (but I don't know what the code is doing) ;) or remove the headier.

File: templates/docs/base/code.md [hunk 16]
```diff
@@ -34,7 +34,7 @@ View example of the base pre block
 View example of the base code block
 </a></div>
 
-### Code Snippets
+### Code snippet
 
 The code snippet pattern should be used to group related code examples, making them easier to find and understand. It should consist of one or, at most, two code blocks with optional headers for their titles.
```
Comment (bartaz) [inline templates/docs/base/code.md:39]: Would be good to add "new" label under Code snippet heading.

File: templates/docs/base/code.md [hunk 17]
```diff
@@ -52,31 +52,27 @@ A code snippet block can include an icon by using the `.p-code-snippet__block--c
 View example of the code snippet
```
Comment (bartaz) [inline templates/docs/base/code.md:52]: Missed that before, there is a typo in class name `.p-code-snippet__block--code`, should be `--icon`.

File: templates/docs/component-status.md [hunk 18]
```diff
@@ -30,11 +30,23 @@ When we add, make significant updates, or deprecate a component we update their
...
+      <td>The new <code>.p-code-snippet</code> component has been added, to be used to display code examples in a number of different formats.</td>
+    </tr>
+    <tr>
+      <th><a href="/docs/examples/patterns/code-copyable">Code copyable</a></th>
       <td><div class="p-label--deprecated">Deprecated</div></td>
       <td>2.22.0</td>
       <td><code>.p-code-copyable</code> has been deprecated and will be removed in the v3.0 release. Please use <code>.p-code-snippet .p-code-snippet__block--icon</code> instead.</td>
     </tr>
+    <tr>
+      <th><a href="/docs/examples/patterns/code-numbered">Code numbered</a></th>
```
Comment (bartaz) [inline templates/docs/component-status.md:45]: Please link to documentation (with #section anchors if needed) rather than examples.

File: scss/_base_code.scss (hunk 14 above)
Comment (sowasred2012) [reply scss/_base_code.scss:5]: I think we should keep this, we position the number column negative left by the width of one character, I think it helps make that clear?

Comment (bartaz) [reply scss/_base_code.scss:5]: Ah, haven't noticed that we do use negative positioning. Sure thing.

**Review by bartaz — APPROVED** [review]
LGTM

File: scss/_base_code.scss (hunk 14 above)
Comment (anthonydillon) [reply scss/_base_code.scss:5]: Thats @lyubomir-popov that is what I was thinking. 

I believe its clearer to simplify the code like this to show its 4 characters width your calculating:
```scss
code-sidebar-width: calc(#{$sph-inner} + 4ch) !default;
```
Over this:
```scss
$digit-width: 1ch;
$code-sidebar-width: calc(#{$sph-inner} + (4 * #{$digit-width})) !default;
```
